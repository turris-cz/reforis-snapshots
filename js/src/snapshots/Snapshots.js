/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useState } from "react";
import PropTypes from "prop-types";

import { API_STATE, Spinner, ErrorMessage, formFieldsSize } from "foris";

import {
    useGetSnapshots,
    useUpdateSnapshotsOnAdd,
    useCreateSnapshot,
    useDeleteSnapshot,
    useUpdateSnapshotsOnDelete,
    useRollbackSnapshot,
    useUpdateSnapshotsOnRollback,
} from "./hooks";
import CreateSnapshotForm from "./CreateSnapshotForm";
import SnapshotsTable from "./SnapshotsTable/SnapshotsTable";

Snapshots.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function Snapshots({ ws }) {
    const [snapshots, setSnapshots] = useState([]);

    const [getState, getSnapshots] = useGetSnapshots(setSnapshots);
    useUpdateSnapshotsOnAdd(ws, getSnapshots);

    const [, createSnapshot] = useCreateSnapshot();

    const [, rollbackSnapshot] = useRollbackSnapshot();
    useUpdateSnapshotsOnRollback(ws, getSnapshots);

    const [, deleteSnapshot] = useDeleteSnapshot();
    useUpdateSnapshotsOnDelete(ws, setSnapshots);

    let componentContent;
    if (getState === API_STATE.INIT || [getState].includes(API_STATE.SENDING)) {
        componentContent = <Spinner />;
    } else if (getState === API_STATE.ERROR) {
        componentContent = <ErrorMessage />;
    } else {
        componentContent = (
            <div className={formFieldsSize}>
                <CreateSnapshotForm createSnapshot={createSnapshot} />
                <h2>{_("Available Snapshots")}</h2>
                <SnapshotsTable
                    snapshots={snapshots}
                    rollbackSnapshot={rollbackSnapshot}
                    deleteSnapshot={deleteSnapshot}
                />
            </div>
        );
    }

    return (
        <>
            <h1>{_("Snapshots")}</h1>
            <p
                dangerouslySetInnerHTML={{
                    __html: _(
                        "This is an addition to Schnapps command-line utility which can provide more <a href='https://docs.turris.cz/geek/schnapps/schnapps/#what-you-can-do-with-schnapps' target='_blank' rel='noopener noreferrer'>advanced options<sup><i class='fas fa-external-link-alt link-outside-icon fa-xs'></i></a>."
                    ),
                }}
            />
            {componentContent}
        </>
    );
}
