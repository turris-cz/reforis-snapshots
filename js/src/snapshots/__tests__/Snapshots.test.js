/*
 * Copyright (C) 2020-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import mockAxios from "jest-mock-axios";
import {
    render,
    wait,
    getAllByText,
    queryByText,
    getByText,
    getByLabelText,
    queryByRole,
    getByRole,
    act,
    fireEvent,
    getByTestId,
} from "foris/testUtils/customTestRender";
import { mockJSONError } from "foris/testUtils/network";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { WebSockets } from "foris";

import Snapshots from "../Snapshots";

const SNAPSHOTS = [
    {
        number: 1,
        type: "single",
        description: "Whatever",
        created: "2020-01-30T10:27:34Z",
        size: "808 kB",
    },
    {
        number: 2,
        type: "rollback",
        description: "Something",
        created: "2020-01-31T10:27:34Z",
        size: "909 kB",
    },
];

function creteSnapshot(number, description) {
    return {
        number,
        type: "single",
        description,
        created: "2020-01-30T10:27:34Z",
        size: "808 kB",
    };
}

describe("<Snapshots />", () => {
    let container;
    let webSockets;

    beforeEach(() => {
        webSockets = new WebSockets();
        ({ container } = render(<Snapshots ws={webSockets} />));
    });

    it("should render spinner", () => {
        expect(container).toMatchSnapshot();
    });

    it("should render table", async () => {
        expect(mockAxios.get).toBeCalledWith(
            "/reforis/snapshots/api/snapshots",
            expect.anything()
        );
        mockAxios.mockResponse({ data: SNAPSHOTS });
        await wait(() => getByText(container, SNAPSHOTS[0].description));
        expect(container).toMatchSnapshot();
    });

    it("should handle GET error", async () => {
        mockJSONError();
        await wait(() =>
            expect(
                getByText(container, "An error occurred while fetching data.")
            ).toBeTruthy()
        );
    });

    it("should display spinner while snapshot is being added", async () => {
        // Prepare table
        mockAxios.mockResponse({ data: SNAPSHOTS });
        // Initially there's no spinner
        await wait(() => expect(queryByRole(container, "status")).toBeNull());

        // Create new snapshot
        act(() =>
            webSockets.dispatch({
                module: "schnapps",
                action: "create",
                data: {},
            })
        );
        // Spinner should appear
        await wait(() => getByRole(container, "status"));
    });

    it("should display spinner when snapshot is being removed", async () => {
        // Prepare table
        mockAxios.mockResponse({ data: SNAPSHOTS });
        // Initially there's no spinner
        await wait(() => expect(queryByRole(container, "status")).toBeNull());

        // Delete device
        fireEvent.click(getAllByText(container, "Delete")[0]);
        // Spinner should appear
        await wait(() => getByRole(container, "status"));
    });

    it("should display spinner while rolling back to snapshot", async () => {
        // Prepare table
        mockAxios.mockResponse({ data: SNAPSHOTS });
        // Initially there's no spinner
        await wait(() => expect(queryByRole(container, "status")).toBeNull());

        // Delete device
        fireEvent.click(getAllByText(container, "Rollback")[0]);
        // Spinner should appear
        await wait(() => getByRole(container, "status"));
    });

    describe("with table prepared", () => {
        beforeEach(async () => {
            // Prepare table
            mockAxios.mockResponse({ data: SNAPSHOTS });
            await wait(() => getByText(container, SNAPSHOTS[0].description));
        });

        it("should refresh table after new snapshot is added", async () => {
            // Create new snapshot
            const newDescription = "Nothing";
            act(() =>
                webSockets.dispatch({
                    module: "schnapps",
                    action: "create",
                    data: {},
                })
            );
            expect(mockAxios.get).toHaveBeenNthCalledWith(
                2,
                "/reforis/snapshots/api/snapshots",
                expect.anything()
            );
            mockAxios.mockResponse({
                data: [...SNAPSHOTS, creteSnapshot(3, newDescription)],
            });
            // New snapshot should appear
            await wait(() => getByText(container, newDescription));
        });

        it("should refresh table after snapshot is removed", async () => {
            // Delete snapshot
            const deletedNumber = SNAPSHOTS[0].number;
            fireEvent.click(getAllByText(container, "Delete")[0]);
            expect(mockAxios.delete).toBeCalledWith(
                `/reforis/snapshots/api/snapshots/${deletedNumber}`,
                expect.anything()
            );
            mockAxios.mockResponse({ data: {} });

            act(() =>
                webSockets.dispatch({
                    module: "schnapps",
                    action: "delete",
                    data: { number: deletedNumber },
                })
            );
            // Device should disappear
            await wait(() =>
                expect(
                    queryByText(container, SNAPSHOTS[0].description)
                ).toBeNull()
            );
        });

        it("should handle error on removal", async () => {
            // Delete device
            fireEvent.click(getAllByText(container, "Delete")[0]);
            // Handle error
            const errorMessage = "API didn't handle this well";
            mockJSONError(errorMessage);
            await wait(() => {
                expect(mockSetAlert).toHaveBeenCalledWith(errorMessage);
            });
        });

        it("should refresh table after rolling back to snapshot", async () => {
            // Rollback to snapshot
            const rollbackToNumber = SNAPSHOTS[0].number;
            fireEvent.click(getAllByText(container, "Rollback")[0]);
            expect(mockAxios.put).toBeCalledWith(
                `/reforis/snapshots/api/snapshots/${rollbackToNumber}/rollback`,
                undefined,
                expect.anything()
            );
            mockAxios.mockResponse({ data: {} });

            act(() =>
                webSockets.dispatch({
                    module: "schnapps",
                    action: "rollback",
                    data: {},
                })
            );

            const newDescription = `Rollback to ${rollbackToNumber}`;
            expect(mockAxios.get).toHaveBeenNthCalledWith(
                2,
                "/reforis/snapshots/api/snapshots",
                expect.anything()
            );
            mockAxios.mockResponse({
                data: [...SNAPSHOTS, creteSnapshot(3, newDescription)],
            });
            // New snapshot should appear
            await wait(() => getByText(container, newDescription));
        });

        it("should handle error on rollback", async () => {
            // Delete device
            fireEvent.click(getAllByText(container, "Rollback")[0]);
            // Handle error
            const errorMessage = "API didn't handle this well";
            mockJSONError(errorMessage);
            await wait(() => {
                expect(mockSetAlert).toHaveBeenCalledWith(errorMessage);
            });
        });

        describe("create snapshot form", () => {
            let descriptionInput;
            let submitButton;

            beforeEach(async () => {
                descriptionInput = getByLabelText(container, "Description");
                submitButton = getByTestId(container, "create-snapshot");
            });

            it("should send request to create new snapshot", async () => {
                // Prepare form
                const description = "Discovery";
                expect(submitButton.disabled).toBe(true);
                fireEvent.change(descriptionInput, {
                    target: { value: description },
                });
                expect(submitButton.disabled).toBe(false);

                // Create new snapshot
                fireEvent.click(submitButton);
                expect(mockAxios.post).toBeCalledWith(
                    "/reforis/snapshots/api/snapshots",
                    { description },
                    expect.anything()
                );
            });

            it("should handle API error on creating snapshot", async () => {
                // Request new snapshot
                fireEvent.change(descriptionInput, {
                    target: { value: "qwe" },
                });
                fireEvent.click(submitButton);

                // Handle error
                const errorMessage = "API didn't handle this well";
                mockJSONError(errorMessage);
                await wait(() => {
                    expect(mockSetAlert).toHaveBeenCalledWith(errorMessage);
                });
            });

            it("should validate new snapshot description", async () => {
                fireEvent.change(descriptionInput, { target: { value: "" } });
                expect(
                    getByText(container, "Description is required.")
                ).toBeDefined();
                expect(submitButton.disabled).toBe(true);
            });
        });
    });
});
