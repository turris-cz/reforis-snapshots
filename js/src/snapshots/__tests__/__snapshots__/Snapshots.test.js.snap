// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Snapshots /> should render spinner 1`] = `
<div>
  <h1>
    Snapshots
  </h1>
  <p>
    This is an addition to Schnapps command-line utility which can provide more 
    <a
      href="https://docs.turris.cz/geek/schnapps/schnapps/#what-you-can-do-with-schnapps"
      rel="noopener noreferrer"
      target="_blank"
    >
      advanced options
      <sup>
        <i
          class="fa"
        />
      </sup>
    </a>
    .
  </p>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;

exports[`<Snapshots /> should render table 1`] = `
<div>
  <h1>
    Snapshots
  </h1>
  <p>
    This is an addition to Schnapps command-line utility which can provide more 
    <a
      href="https://docs.turris.cz/geek/schnapps/schnapps/#what-you-can-do-with-schnapps"
      rel="noopener noreferrer"
      target="_blank"
    >
      advanced options
      <sup>
        <i
          class="fa"
        />
      </sup>
    </a>
    .
  </p>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <form
      class="mb-3 mb-md-0"
    >
      <h2>
        Create Snapshot
      </h2>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="1"
        >
          Description
        </label>
        <div
          class="input-group"
        >
          <input
            class="form-control"
            id="1"
            maxlength="50"
            type="text"
            value=""
          />
        </div>
        
      </div>
      <div
        class="text-end"
      >
        <button
          class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
          data-testid="create-snapshot"
          disabled=""
          type="submit"
        >
          <span>
            Create
          </span>
        </button>
      </div>
    </form>
  </div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Available Snapshots
    </h2>
    <div>
      <div
        class="d-flex justify-content-between align-items-center"
      >
        <div
          class="mb-3"
        >
          <div
            class="input-group"
          >
            <input
              class="form-control me-3"
              id="2"
              placeholder="Searchâ€¦"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="dropdown mb-3"
        >
          <button
            class="btn btn btn-outline-secondary dropdown-toggle d-inline-flex justify-content-center align-items-center"
            data-bs-toggle="dropdown"
            type="button"
          >
            <span>
              Columns
            </span>
          </button>
          <ul
            class="dropdown-menu dropdown-menu-end"
          >
            <li>
              <button
                class="dropdown-item d-flex align-items-center"
                style="padding-left: 2rem;"
                type="button"
              >
                <i
                  class="fa"
                />
                <span>
                  #
                </span>
              </button>
            </li>
            <li>
              <button
                class="dropdown-item d-flex align-items-center"
                style="padding-left: 2rem;"
                type="button"
              >
                <i
                  class="fa"
                />
                <span>
                  Description
                </span>
              </button>
            </li>
            <li>
              <button
                class="dropdown-item d-flex align-items-center"
                style="padding-left: 2rem;"
                type="button"
              >
                <i
                  class="fa"
                />
                <span>
                  Created at
                </span>
              </button>
            </li>
            <li>
              <button
                class="dropdown-item d-flex align-items-center"
                style="padding-left: 2rem;"
                type="button"
              >
                <i
                  class="fa"
                />
                <span>
                  Size
                </span>
              </button>
            </li>
            <li>
              <button
                class="dropdown-item d-flex align-items-center"
                style="padding-left: 2rem;"
                type="button"
              >
                <i
                  class="fa"
                />
                <span>
                  Actions
                </span>
              </button>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="table-responsive"
      >
        <table
          class="table table-hover text-nowrap"
        >
          <thead
            class="table-light"
          >
            <tr
              role="row"
            >
              <th
                colspan="1"
              >
                <button
                  class="btn btn-link text-decoration-none text-reset fw-bold p-0 d-flex align-items-center
                                                    d-flex align-items-center
                                                "
                  title="Sort descending"
                  type="button"
                >
                  #
                </button>
              </th>
              <th
                colspan="1"
              >
                <button
                  class="btn btn-link text-decoration-none text-reset fw-bold p-0 d-flex align-items-center
                                                    d-flex align-items-center
                                                "
                  title="Sort ascending"
                  type="button"
                >
                  Description
                </button>
              </th>
              <th
                colspan="1"
              >
                <button
                  class="btn btn-link text-decoration-none text-reset fw-bold p-0 d-flex align-items-center
                                                    d-flex align-items-center
                                                "
                  title="Sort ascending"
                  type="button"
                >
                  Created at
                </button>
              </th>
              <th
                colspan="1"
              >
                <button
                  class="btn btn-link text-decoration-none text-reset fw-bold p-0 d-flex align-items-center
                                                    d-flex align-items-center
                                                "
                  title="Sort ascending"
                  type="button"
                >
                  Size
                </button>
              </th>
              <th
                colspan="1"
              >
                <div
                  aria-hidden="true"
                  class="d-none"
                >
                  Actions
                </div>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="align-middle"
            >
              <td>
                2
              </td>
              <td>
                Something
              </td>
              <td>
                1/31/2020 10:27 AM
              </td>
              <td>
                909 kB
              </td>
              <td
                class="text-end"
              >
                <div
                  class="dropdown position-static"
                  data-testid="three-dots-menu"
                >
                  <button
                    aria-expanded="false"
                    class="btn btn-sm btn-link text-body d-inline-flex justify-content-center align-items-center"
                    data-bs-toggle="dropdown"
                    type="button"
                  >
                    <span>
                      <i
                        class="fa"
                      />
                    </span>
                  </button>
                  <ul
                    class="dropdown-menu"
                  >
                    <li>
                      <button
                        class="dropdown-item"
                        data-testid="three-dots-menu-download"
                        type="button"
                      >
                        <i
                          class="fa"
                        />
                        Download
                      </button>
                    </li>
                    <li>
                      <button
                        class="dropdown-item"
                        data-testid="three-dots-menu-rollback"
                        type="button"
                      >
                        <i
                          class="fa"
                        />
                        Rollback
                      </button>
                    </li>
                    <li>
                      <button
                        class="dropdown-item"
                        data-testid="three-dots-menu-delete"
                        type="button"
                      >
                        <i
                          class="fa"
                        />
                        Delete
                      </button>
                    </li>
                  </ul>
                </div>
              </td>
            </tr>
            <tr
              class="align-middle"
            >
              <td>
                1
              </td>
              <td>
                Whatever
              </td>
              <td>
                1/30/2020 10:27 AM
              </td>
              <td>
                808 kB
              </td>
              <td
                class="text-end"
              >
                <div
                  class="dropdown position-static"
                  data-testid="three-dots-menu"
                >
                  <button
                    aria-expanded="false"
                    class="btn btn-sm btn-link text-body d-inline-flex justify-content-center align-items-center"
                    data-bs-toggle="dropdown"
                    type="button"
                  >
                    <span>
                      <i
                        class="fa"
                      />
                    </span>
                  </button>
                  <ul
                    class="dropdown-menu"
                  >
                    <li>
                      <button
                        class="dropdown-item"
                        data-testid="three-dots-menu-download"
                        type="button"
                      >
                        <i
                          class="fa"
                        />
                        Download
                      </button>
                    </li>
                    <li>
                      <button
                        class="dropdown-item"
                        data-testid="three-dots-menu-rollback"
                        type="button"
                      >
                        <i
                          class="fa"
                        />
                        Rollback
                      </button>
                    </li>
                    <li>
                      <button
                        class="dropdown-item"
                        data-testid="three-dots-menu-delete"
                        type="button"
                      >
                        <i
                          class="fa"
                        />
                        Delete
                      </button>
                    </li>
                  </ul>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
